### PR TITLE
Add simple proxy

### PR DESCRIPTION
**Pull request by AVierwind**
Add's a simple proxy to the project. Proxy should be configured by adding a proxyConfig to your config file. For details view README.md #30

### CODE REVIEW

**Comment by simonh1000** [comment]
This looks rather cool. Let me take some time to review it. But could you first add some extra comments about what the use case is for the option - i just want to make sure we are on the same page

**Comment by AVierwind** [comment]
I have a couple of apps that will do an API request to they own host, in production cases the webserver wil proxy those requests to the web api's. This option wil mimic that behavior to make it possible to run those apps on this server. example: When my app starts it will do a request to /api-proxy/get-user-data which is proxied to the actual api. This simple proxy can be used to proxy those request to a mockserver so i can also use this server to spin up those apps

File: lib/proxy.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                `Proxying request ${req.method}, ${req.url} to ${protocol}://${config.target}`
+            );
+            const proxy = httpProxy.createProxyServer({});
+            proxy.web(req, res, {
+                target: `${protocol}://${config.target}`,
+            });
+        }
+    });
+
+    return proxyHit;
```
Comment (simonh1000) [inline lib/proxy.js:41]: this is going to return the value of line 28 for the last config. But you want it to return whether the proxy was found in any of the configs

File: lib/proxy.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            throw new Error("target is missing or empty in your proxyConfig");
+        }
+
+        proxyHit = config.forward.some((url) => req.url.indexOf(url) !== -1);
+        if (proxyHit) {
+            const protocol = useHttps ? "https" : "http";
+            console.log(
+                `Proxying request ${req.method}, ${req.url} to ${protocol}://${config.target}`
+            );
+            const proxy = httpProxy.createProxyServer({});
```
Comment (simonh1000) [inline lib/proxy.js:34]: this is going to create a proxyServer for the same `forward` every time we navigate to a new route? Isn't that a bit wasteful?

File: README.md [hunk 3]
```diff
@@ -111,6 +111,41 @@ module.exports = (argv) => {
...
 
+## Http proxy
+
+#### Enabling proxy
+
+The server contains a simple http proxy. 
+The proxy must be configured in the config file.
+To enable this proxy:
+
+```sh
```
Comment (simonh1000) [inline README.md:122]: I don't think we need the flag - this feature can only be implemented using a config file and either the user has added `proxy` options or they haven't

File: README.md [hunk 4]
```diff
@@ -111,6 +111,41 @@ module.exports = (argv) => {
...
+#### configuring proxy
+
+To configure the proxy add a proxyConfig object to your config file. 
+The proxyConfig should be an array of configs with two properties: a forward property which must be a string array listing url parts which should trigger the proxy, and a target property which should define the target to proxy to.
+
+simple example:
+
+```js
+module.exports = {
+    proxyConfig: [
```
Comment (simonh1000) [inline README.md:135]: as we are in a config file anyway, how about just calling this `proxy`

File: lib/proxy.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * Simple Http Proxy, proxies request where url is contained in config to target from config
+ *
+ * @param req
+ * @param res
+ * @param {object} configs - proxyConfig from config file
+ * @param {boolean} useHttps - will determine protocol to use
+ * @return {boolean} - will return whether request is being proxied
+ */
+function proxyHandler(req, res, configs, useHttps) {
+    if (!configs || configs.length === 0) {
```
Comment (simonh1000) [inline lib/proxy.js:13]: after this merged, I will create a config validation function and move this code in there

**Review by simonh1000 — CHANGES_REQUESTED** [review]
I haven't tested it but can see how it will work. But I wonder about the way you are creating so many proxies over time

File: lib/proxy.js (hunk 1 above)
Comment (AVierwind) [reply lib/proxy.js:54]: changes it so nothing in the foreach loop wil happen when it has already fount a hit

File: lib/proxy.js (hunk 2 above)
Comment (AVierwind) [reply lib/proxy.js]: changes it so proxy will be created only once

File: README.md (hunk 3 above)
Comment (AVierwind) [reply README.md:122]: removed the flag, added check for the presence of proxy config before calling the proxy

File: README.md (hunk 4 above)
Comment (AVierwind) [reply README.md]: done

**Comment by simonh1000** [comment]
Hi there. I tried it out and hit a small issue. I want the server to run without having to worry about https, but if I don't use --https then your code uses `http` when it forwards requests, which will generally fail. I looked at how webpack do it and they provide an option to add `protocol: 'http:'` in their config. Do you think it would make sense to have it here.

(the problem is exacerbated by the fact that the `--https` option no longer seems to create assets that chrome will accept - not your problem, but it makes using your option harder. How do you use it?)

**Comment by AVierwind** [comment]
> Hi there. I tried it out and hit a small issue. I want the server to run without having to worry about https, but if I don't use --https then your code uses `http` when it forwards requests, which will generally fail. I looked at how webpack do it and they provide an option to add `protocol: 'http:'` in their config. Do you think it would make sense to have it here.
> 
> (the problem is exacerbated by the fact that the `--https` option no longer seems to create assets that chrome will accept - not your problem, but it makes using your option harder. How do you use it?)

I Just pushed a change, which adds a protocol option to the config instead of listening to the --htttps option. When absent https will be the default

**Comment by simonh1000** [comment]
I tried to test again - see https://github.com/simonh1000/angular-http-server/pull/48 - and still cant get it to work. I've run into a cloudflare issue `Direct IP access not allowed`.  My guess is that your branch is tuned to your use case, but I've learnt in my ftp packager that in the wide world of network communications all sorts of weirdness can happen.

I'm going to try to build a server on my machine and proxy to that.

one change that I tried that got me past the first error but not the second was 

```
httpProxy.createProxyServer({
      secure: false,
});
``` 

Any views on this?

File: lib/proxy.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                throw new Error(
+                    "target is missing or empty in your proxyConfig"
+                );
+            }
+
+            proxyHit = config.forward.some(
+                (url) => req.url.indexOf(url) !== -1
+            );
+            if (proxyHit) {
+                const protocol = config.protocol ? "https" : config.protocol;
```
Comment (simonh1000) [inline lib/proxy.js:38]: the then and else returns are the wrong way around

**Comment by simonh1000** [comment]
I have now got it to work on my machine with a reasonable test case. There is the bug above but then we can merge

File: lib/proxy.js (hunk 6 above)
Comment (AVierwind) [reply lib/proxy.js]: Oh wow! fixed it

**Comment by simonh1000** [comment]
Cool, will merge and release a new version at the weekend

**Pull request merged**